File: javascript/src/bursar.ts
import {
  BillingService as BillingEventService,
  type BillingServiceOptions,
} from "./billing/billing-service.js";
import type { BillingStore } from "./billing/billing-store.js";
import {
  CreditsService as CreditsServiceImpl,
  type CreditsServiceOptions,
} from "./credits/service.js";
import type { CreditStore } from "./credits/store.js";
import type { CreditEventEmitter } from "./credits/events.js";
import type { BillingEvent, BillingEventResult } from "./billing/types/index.js";
import type { BillingCapability, BillingEventSink } from "./billing/contracts.js";
import { CommerceService as CommerceServiceImpl } from "./commerce/service.js";
import { CommerceNotConfiguredError } from "./commerce/errors.js";
import type { CommerceOptions } from "./commerce/types.js";
import { loadConfigFromDict } from "./config.js";
import type { BursarConfigData, CatalogRollout, ParsedBursarConfig } from "./config/types.js";
import { CapabilityNotConfiguredError, ConfigError, CatalogNotLoadedError } from "./errors.js";
import { projectPublicCatalog, type PublicCatalog } from "./catalog.js";
import type { CreditMetadata, GrantProgramAwardResult } from "./credits/types/index.js";
import { retryBursarOperation } from "./retry.js";
import { z } from "zod";
export type { BillingCapability, BillingEventSink } from "./billing/contracts.js";
export type { CommerceOptions } from "./commerce/types.js";

/**
 * Stable public credit capability exposed by {@link Bursar}.
 *
 * The explicit method list prevents implementation helpers added to
 * `CreditsServiceImpl` from silently becoming public API.
 */
export type CreditsService = Pick<
  CreditsServiceImpl,
  | "getUserPlan"
  | "checkFeature"
  | "getQuotaState"
  | "listQuotaEvents"
  | "startPlanMigration"
  | "migratePlanBatch"
  | "revokeCreditsByEntryType"
  | "executeGrantProgram"
  | "getLedgerEntry"
  | "getAvailable"
  | "aggregateStats"
  | "spendByUser"
  | "spendByModel"
  | "listLedgerEntries"
  | "listUsageEntries"
  | "listUsageCharges"
  | "topUsers"
  | "dailySpend"
  | "setUserPlan"
  | "unsetUserPlan"
  | "getBalance"
  | "addCredits"
  | "deductCredits"
  | "grantSubscriptionCycle"
  | "reserve"
  | "settle"
  | "release"
  | "renew"
  | "canAfford"
  | "getBucketBalances"
  | "checkAllowance"
  | "runBilled"
  | "beginBilledOperation"
  | "deduct"
  | "recordUsage"
  | "deductFlatJob"
  | "refundCredits"
  | "deductTeam"
  | "sweepExpiredCredits"
>;

type CatalogCreditsService = Pick<
  CreditsServiceImpl,
  | "getActiveCatalog"
  | "loadCatalogFromStore"
  | "refreshCatalogIfStale"
  | "invalidateCatalog"
  | "publishAndActivateCatalog"
  | "publishCatalogDraft"
  | "activateCatalogRevision"
  | "pricingEngine"
  | "setPlanRevisionPin"
  | "applyDuePlanChanges"
>;

interface BaseBursarOptions {
  creditStore: CreditStore;
  creditsOptions?: CreditsServiceOptions;
  emitter?: CreditEventEmitter;
}

/** Credits-only construction rejects billing options that would otherwise be ignored. */
export interface CreditsOnlyBursarOptions extends BaseBursarOptions {
  billingStore?: undefined;
  billingOptions?: never;
  commerceOptions?: never;
}

/** Billing construction owns provisioning and may optionally enable commerce. */
export interface BillingBursarOptions extends BaseBursarOptions {
  billingStore: BillingStore;
  billingOptions?: Omit<BillingServiceOptions, "provisioning">;
  commerceOptions?: CommerceOptions;
}

/** Options for constructing the single application-facing Bursar service. */
export type BursarOptions = CreditsOnlyBursarOptions | BillingBursarOptions;

/** Catalog operations. Configuration writes live here, never in BillingService. */
export class CatalogService {
  constructor(private readonly credits: CatalogCreditsService) {}

  /** Return the active persisted catalog revision, if one exists. */
  getActive() {
    return this.credits.getActiveCatalog();
  }

  /** Whether this process currently has a pricing engine loaded. */
  get isLoaded(): boolean {
    return this.credits.pricingEngine !== null;
  }

  /** Load the active persisted catalog into this process. */
  load(): Promise<void> {
    return this.credits.loadCatalogFromStore();
  }

  /** Block until a stale in-process catalog has been refreshed. */
  refresh(): Promise<void> {
    return this.credits.refreshCatalogIfStale();
  }

  /** Force the next refresh to reload the active catalog. */
  invalidate(): void {
    this.credits.invalidateCatalog();
  }

  async getConfig(): Promise<ParsedBursarConfig> {
    const active = await this.getActive();
    if (!active) throw new CatalogNotLoadedError("No active Bursar catalog is available");
    return loadConfigFromDict(active.config);
  }

  async publicView(): Promise<PublicCatalog> {
    return projectPublicCatalog(await this.getConfig());
  }

  publishDraft(config: BursarConfigData, label?: string | null): Promise<string> {
    return this.credits.publishCatalogDraft(config, label);
  }

  activate(version: number, rollout?: CatalogRollout | null): Promise<string> {
    return this.credits.activateCatalogRevision(version, rollout);
  }

  publishAndActivate(
    config: BursarConfigData,
    label?: string | null,
    rollout?: CatalogRollout | null,
  ): Promise<string> {
    return this.credits.publishAndActivateCatalog(config, label, rollout);
  }

  /** Pin or unpin one current assignment from automatic catalog rollout. */
  setRevisionPin(userId: string, pinned: boolean): Promise<boolean> {
    return this.credits.setPlanRevisionPin(userId, pinned);
  }

  /** Apply one bounded batch of renewal-effective plan changes. */
  applyDueChanges(limit = 100): Promise<number> {
    return this.credits.applyDuePlanChanges(limit);
  }
}

export interface AccountCreatedInput {
  accountId: string;
  eventKey: string;
  region?: string | null;
  metadata?: CreditMetadata | null;
}

export interface AccountCreatedResult {
  accountId: string;
  planKey: string;
  planAssigned: boolean;
  grants: GrantProgramAwardResult[];
}

/** Generic financial lifecycle operations for SaaS accounts. */
export class AccountService {
  constructor(
    private readonly credits: CreditsService,
    private readonly catalog: CatalogService,
  ) {}

  async onAccountCreated(input: AccountCreatedInput): Promise<AccountCreatedResult> {
    if (!z.string().trim().min(1).safeParse(input.accountId).success) {
      throw new TypeError("accountId must be a non-empty string");
    }
    if (!z.string().trim().min(1).safeParse(input.eventKey).success) {
      throw new TypeError("eventKey must be a non-empty string");
    }
    const config = await retryBursarOperation(() => this.catalog.getConfig());
    const planKey = config.catalog.defaultPlan;
    if (!planKey) throw new ConfigError("The active catalog has no default account plan");

    const current = await retryBursarOperation(() => this.credits.getUserPlan(input.accountId));
    const planAssigned = current.planKey == null;
    if (planAssigned) {
      await retryBursarOperation(() => this.credits.setUserPlan(input.accountId, planKey));
    }

    const grants: GrantProgramAwardResult[] = [];
    for (const [programKey, program] of Object.entries(config.credits.grantPrograms).sort(
      ([a], [b]) => a.localeCompare(b),
    )) {
      if (program.trigger !== "account_created") continue;
      grants.push(
        ...(await retryBursarOperation(() =>
          this.credits.executeGrantProgram({
            trigger: "account_created",
            programKey,
            subjectId: input.accountId,
            eventKey: input.eventKey,
            region: input.region,
            metadata: input.metadata,
          }),
        )),
      );
    }
    return {
      accountId: input.accountId,
      planKey: current.planKey ?? planKey,
      planAssigned,
      grants,
    };
  }
}

/**
 * The application-facing Bursar boundary.
 *
 * Credit and billing services are deliberately created together so consumers
 * cannot accidentally construct unrelated services for the same account
 * lifecycle. Provider and application integrations use this facade rather
 * than constructing lifecycle services independently.
 */
export class Bursar implements BillingEventSink {
  readonly credits: CreditsService;
  readonly billing: BillingCapability | null;
  readonly commerce: CommerceServiceImpl | null;
  readonly catalog: CatalogService;
  readonly accounts: AccountService;

  constructor(options: BursarOptions) {
    if (!z.object({}).safeParse(options).success) {
      throw new TypeError("Bursar options are required");
    }
    if (options.billingStore === null) {
      throw new TypeError("billingStore must be a BillingStore when provided");
    }
    for (const optionName of [
      "creditsOptions",
      "billingOptions",
      "commerceOptions",
      "emitter",
    ] as const) {
      if (options[optionName] === null) {
        throw new TypeError(`${optionName} must not be null`);
      }
    }
    if (
      options.billingStore === undefined &&
      (options.billingOptions !== undefined || options.commerceOptions !== undefined)
    ) {
      throw new TypeError("billingOptions and commerceOptions require billingStore");
    }
    if (options.billingOptions && "provisioning" in options.billingOptions) {
      throw new TypeError("billingOptions.provisioning is owned by Bursar");
    }
    const environments = [
      ["creditStore", options.creditStore.providerEnvironment],
      ["billingStore", options.billingStore?.providerEnvironment],
      ["commerceOptions", options.commerceOptions?.providerEnvironment],
    ].filter((entry): entry is [string, NonNullable<(typeof entry)[1]>] => entry[1] !== undefined);
    if (new Set(environments.map(([, environment]) => environment)).size > 1) {
      throw new TypeError(
        `Bursar provider environments must match: ${environments
          .map(([name, environment]) => `${name}=${environment}`)
          .join(", ")}`,
      );
    }
    const credits = new CreditsServiceImpl(
      options.creditStore,
      undefined,
      options.emitter ?? undefined,
      options.creditsOptions ?? undefined,
    );
    this.credits = credits;
    this.catalog = new CatalogService(credits);
    this.accounts = new AccountService(credits, this.catalog);

    this.billing = options.billingStore
      ? new BillingEventService(options.billingStore, {
          ...(options.billingOptions ?? {}),
          provisioning: credits,
        })
      : null;
    this.commerce =
      this.billing && options.commerceOptions
        ? new CommerceServiceImpl(this.billing, credits, this, options.commerceOptions)
        : null;
    if (this.commerce) {
      credits.addPostDeductionHook(async ({ userId }) => {
        await this.commerce!.autoRecharge.processIfNeeded({ accountId: userId });
      });
    }
  }

  /** Load the active catalog into the pricing engine. */
  async loadCatalog(): Promise<void> {
    await this.catalog.load();
  }

  /** Return billing or raise the SDK's typed configuration error. */
  requireBilling(): BillingCapability {
    if (!this.billing) {
      throw new CapabilityNotConfiguredError("billing");
    }
    return this.billing;
  }

  /** Return commerce or raise the SDK's typed configuration error. */
  requireCommerce(): CommerceServiceImpl {
    if (!this.commerce) {
      throw new CommerceNotConfiguredError("Bursar commerce capability is not configured");
    }
    return this.commerce;
  }

  /**
   * Ingest a normalized provider event through the facade-owned billing
   * lifecycle. Providers must not depend on BillingService directly.
   */
  async ingestBillingEvent(event: BillingEvent): Promise<BillingEventResult> {
    return this.requireBilling().ingestBillingEvent(event);
  }
}

File: golang/bursar.go
// Copyright 2026 Zonastery
// SPDX-License-Identifier: AGPL-3.0-only

package bursar

import (
	"context"
	"fmt"
	"sort"
	"time"
)

// BillingServiceOptions configures application-owned handlers around Bursar's
// durable billing event lifecycle. Financial provisioning belongs in these
// handlers or a BillingStore implementation; it is never inferred from an
// unverified provider webhook.
type BillingServiceOptions struct {
	Handlers                    map[BillingEventType]BillingEventHandler
	DefaultHandler              BillingEventHandler
	Provisioning                BillingProvisioningPort
	AutoSelectEntitlementSource *bool
	PastDueGracePeriod          *time.Duration
	TerminalPlanKey             string
}

// Options constructs Bursar's single application-facing facade. CreditStore
// is always required. Billing and commerce are optional capabilities, but
// commerce requires billing and a durable checkout-intent store.
type Options struct {
	CreditStore     CreditStore
	CreditsOptions  CreditsServiceOptions
	Emitter         CreditEventSink
	BillingStore    BillingStore
	BillingOptions  *BillingServiceOptions
	CommerceOptions *CommerceOptions
}

// Bursar groups a tenant's catalog, credits, account lifecycle, billing, and
// optional commerce capabilities. Applications should construct this facade
// once per tenant-bound store instead of wiring unrelated service instances.
type Bursar struct {
	Credits  *CreditsService
	Catalog  *CatalogService
	Accounts *AccountService
	Billing  *BillingService
	Commerce *CommerceService
}

// New constructs the application-facing Bursar facade.
func New(options Options) (*Bursar, error) {
	if options.CreditStore == nil {
		return nil, NewError("Bursar requires a credit store", ErrorOptions{Code: ErrorCodeConfig, Category: ErrorCategoryInvalidRequest})
	}
	if options.BillingStore == nil && (options.BillingOptions != nil || options.CommerceOptions != nil) {
		return nil, NewError("billing and commerce options require a billing store", ErrorOptions{Code: ErrorCodeConfig, Category: ErrorCategoryInvalidRequest})
	}
	if options.Emitter != nil && options.CreditsOptions.EventSink != nil {
		return nil, NewError("configure a credit event sink with either Emitter or CreditsOptions.EventSink", ErrorOptions{Code: ErrorCodeConfig, Category: ErrorCategoryInvalidRequest})
	}
	if err := validateFacadeEnvironments(options); err != nil {
		return nil, err
	}

	creditsOptions := options.CreditsOptions
	if options.Emitter != nil {
		creditsOptions.EventSink = options.Emitter
	}
	credits, err := NewCreditsService(options.CreditStore, creditsOptions)
	if err != nil {
		return nil, err
	}
	catalog := credits.Catalog()
	accounts, err := NewAccountService(credits, catalog)
	if err != nil {
		return nil, err
	}
	sdk := &Bursar{Credits: credits, Catalog: catalog, Accounts: accounts}
	if options.BillingStore == nil {
		return sdk, nil
	}
	var billing *BillingService
	if options.BillingOptions == nil {
		billing, err = NewBillingService(options.BillingStore)
	} else {
		billing, err = NewBillingService(options.BillingStore, *options.BillingOptions)
	}
	if err != nil {
		return nil, err
	}
	if billing.provisioning == nil {
		billing.provisioning = credits
	}
	autoRechargeStore, _ := options.BillingStore.(AutoRechargeStore)
	autoRechargeOptions := AutoRechargeServiceOptions{}
	if options.CommerceOptions != nil {
		if options.CommerceOptions.AutoRechargeStore != nil {
			autoRechargeStore = options.CommerceOptions.AutoRechargeStore
		}
		autoRechargeOptions = options.CommerceOptions.AutoRechargeOptions
	}
	if autoRechargeStore != nil {
		billing.AutoRecharge, err = NewAutoRechargeService(catalog, autoRechargeStore, autoRechargeOptions)
		if err != nil {
			return nil, err
		}
	}
	sdk.Billing = billing
	if options.CommerceOptions == nil {
		return sdk, nil
	}
	commerceOptions := *options.CommerceOptions
	if commerceOptions.Store == nil {
		if store, ok := options.BillingStore.(CommerceStore); ok {
			commerceOptions.Store = store
		}
	}
	if commerceOptions.StateStore == nil {
		if store, ok := options.BillingStore.(CommerceStateStore); ok {
			commerceOptions.StateStore = store
		}
	}
	if commerceOptions.AutoRechargeStore == nil {
		if store, ok := options.BillingStore.(AutoRechargeStore); ok {
			commerceOptions.AutoRechargeStore = store
		}
	}
	commerce, err := NewCommerceService(billing, catalog, credits, commerceOptions)
	if err != nil {
		return nil, err
	}
	sdk.Commerce = commerce
	return sdk, nil
}

// LoadCatalog loads the active persisted catalog into this process's exact
// pricing engine. It must succeed before callers use Catalog.Engine.
func (b *Bursar) LoadCatalog(ctx context.Context) error {
	if b == nil || b.Catalog == nil {
		return NewError("Bursar catalog capability is not configured", ErrorOptions{Code: ErrorCodeCatalogNotLoaded, Category: ErrorCategoryUnavailable})
	}
	return b.Catalog.Load(ctx)
}

// RequireBilling returns billing or a typed capability error.
func (b *Bursar) RequireBilling() (*BillingService, error) {
	if b == nil || b.Billing == nil {
		return nil, NewError("Bursar billing capability is not configured", ErrorOptions{Code: ErrorCodeCapabilityNotConfigured, Category: ErrorCategoryInvalidRequest})
	}
	return b.Billing, nil
}

// RequireCommerce returns commerce or a typed capability error.
func (b *Bursar) RequireCommerce() (*CommerceService, error) {
	if b == nil || b.Commerce == nil {
		return nil, NewError("Bursar commerce capability is not configured", ErrorOptions{Code: ErrorCodeCommerceNotConfigured, Category: ErrorCategoryInvalidRequest})
	}
	return b.Commerce, nil
}

// IngestBillingEvent feeds a previously verified, normalized provider event
// through facade-owned billing event claiming and completion.
func (b *Bursar) IngestBillingEvent(ctx context.Context, event BillingEvent) (BillingEventResult, error) {
	billing, err := b.RequireBilling()
	if err != nil {
		return BillingEventResult{}, err
	}
	return billing.Ingest(ctx, event)
}

// Close releases credit-store resources. Billing stores that own separate
// resources should expose and close their own lifecycle in application setup.
func (b *Bursar) Close() error {
	if b == nil {
		return nil
	}
	if b.Commerce != nil {
		b.Commerce.Close()
	}
	if b.Credits == nil {
		return nil
	}
	return b.Credits.Close()
}

type providerEnvironmentSource interface {
	ProviderEnvironment() ProviderEnvironment
}

func validateFacadeEnvironments(options Options) error {
	environments := map[string]ProviderEnvironment{}
	if source, ok := options.CreditStore.(providerEnvironmentSource); ok {
		if environment := source.ProviderEnvironment(); environment != "" {
			environments["credit_store"] = environment
		}
	}
	if options.BillingStore != nil {
		environments["billing_store"] = options.BillingStore.ProviderEnvironment()
	}
	if options.CommerceOptions != nil && options.CommerceOptions.Providers != nil {
		if environment := options.CommerceOptions.Providers.Environment(); environment != "" {
			environments["commerce_providers"] = environment
		}
	}
	if len(environments) < 2 {
		return nil
	}
	values := make([]ProviderEnvironment, 0, len(environments))
	for _, environment := range environments {
		if err := environment.Validate(); err != nil {
			return err
		}
		values = append(values, environment)
	}
	for _, environment := range values[1:] {
		if environment != values[0] {
			keys := make([]string, 0, len(environments))
			for key := range environments {
				keys = append(keys, key)
			}
			sort.Strings(keys)
			parts := make([]string, 0, len(keys))
			for _, key := range keys {
				parts = append(parts, fmt.Sprintf("%s=%s", key, environments[key]))
			}
			return NewError("Bursar provider environments must match: "+joinStrings(parts, ", "), ErrorOptions{Code: ErrorCodeConfig, Category: ErrorCategoryInvalidRequest})
		}
	}
	return nil
}

func applyBillingOptions(service *BillingService, options *BillingServiceOptions) error {
	if options == nil {
		return nil
	}
	keys := make([]string, 0, len(options.Handlers))
	for eventType := range options.Handlers {
		keys = append(keys, string(eventType))
	}
	sort.Strings(keys)
	for _, key := range keys {
		eventType := BillingEventType(key)
		if err := service.On(eventType, options.Handlers[eventType]); err != nil {
			return err
		}
	}
	service.SetDefaultHandler(options.DefaultHandler)
	if options.Provisioning != nil {
		service.provisioning = options.Provisioning
	}
	service.autoSelectEntitlementSource = true
	if options.AutoSelectEntitlementSource != nil {
		service.autoSelectEntitlementSource = *options.AutoSelectEntitlementSource
	}
	if options.PastDueGracePeriod != nil && *options.PastDueGracePeriod < 0 {
		return NewError("billing past-due grace period must be non-negative", ErrorOptions{Code: ErrorCodeConfig, Category: ErrorCategoryInvalidRequest})
	}
	if options.PastDueGracePeriod != nil {
		service.pastDueGracePeriod = *options.PastDueGracePeriod
	}
	service.terminalPlanKey = options.TerminalPlanKey
	return nil
}

func joinStrings(values []string, separator string) string {
	if len(values) == 0 {
		return ""
	}
	result := values[0]
	for _, value := range values[1:] {
		result += separator + value
	}
	return result
}
